Blazor - rendering metrics and tracing

pavelsavara opened the pull request:
### Blazor metrics
- new meter `Microsoft.AspNetCore.Components`
`aspnetcore.components.navigation` - Total number of route changes.
`aspnetcore.components.event_handler` - Duration of processing browser event including business logic.

- new meter `Microsoft.AspNetCore.Components.Lifecycle`
`aspnetcore.components.update_parameters` - Duration of processing component parameters including business logic.
`aspnetcore.components.render_batch` - Duration of rendering batch including browser round-trip.

- meter `Microsoft.AspNetCore.Components.Server.Circuits`
`aspnetcore.components.circuit.active` - Number of active circuits in memory.
`aspnetcore.components.circuit.connected` - Number of circuits connected to client.
`aspnetcore.components.circuit.duration` - Duration of circuit lifetime and their total count.

### Blazor activity tracing
- new activity source `Microsoft.AspNetCore.Components`
- `Microsoft.AspNetCore.Components.CircuitStart`: `Circuit {circuitId}`
    - tags: `aspnetcore.components.circuit.id`
    - links: HTTP activity
- `Microsoft.AspNetCore.Components.RouteChange`: `Route {route} -> {componentType}`
    - tags: `aspnetcore.components.circuit.id`, `aspnetcore.components.type`, `aspnetcore.components.route`
    - links: HTTP trace, circuit trace
- `Microsoft.AspNetCore.Components.HandleEvent`: `Event {attributeName} -> {componentType}.{methodName}`
    - tags: `aspnetcore.components.circuit.id`, `aspnetcore.components.type`, `aspnetcore.components.method`, `aspnetcore.components.attribute.name`, `error.type`
    - links: HTTP trace, circuit trace, router trace

![image](https://github.com/user-attachments/assets/e301a6d2-6adb-451b-a720-43b0a0681451)

![image](https://github.com/user-attachments/assets/392875ba-bb3e-4860-858e-5d9dbc1a135c)

![image](https://github.com/user-attachments/assets/d2ea09e0-6891-474e-a0b8-008ac661422f)


```cs
builder.Services.ConfigureOpenTelemetryMeterProvider(meterProvider =>
{
    meterProvider.AddMeter("Microsoft.AspNetCore.Components");
    meterProvider.AddMeter("Microsoft.AspNetCore.Components.Server.Circuits");
});
builder.Services.ConfigureOpenTelemetryTracerProvider(tracerProvider =>
{
    tracerProvider.AddSource("Microsoft.AspNetCore.Components");
    //tracerProvider.AddSource("Microsoft.AspNetCore.SignalR.Server");
});
```

### Feedback

- always expect the `IMeterFactory` to be available in DI
- make buckets different size [per feedback](https://github.com/dotnet/aspnetcore/pull/61516#discussion_r2052263412)
- optimize work with tags
- make metrics in seconds [per feedback](https://github.com/dotnet/aspnetcore/pull/61516#discussion_r2052340016)
- unit test [delay feedback](https://github.com/dotnet/aspnetcore/pull/61516#discussion_r2052266880)

TODO - Metrics need to be documented at https://learn.microsoft.com/en-us/aspnet/core/log-mon/metrics/built-in

Out of scope
- WASM


Contributes to https://github.com/dotnet/aspnetcore/issues/53613
Contributes to https://github.com/dotnet/aspnetcore/issues/29846
Feedback for https://github.com/dotnet/aspnetcore/pull/61516
Related https://github.com/open-telemetry/semantic-conventions/pull/2235

pavelsavara committed:
rebase

pavelsavara committed:
Merge branch 'main' into blazor_metrics_feedback

# Conflicts:
#	src/Components/Components/src/PublicAPI.Unshipped.txt

pavelsavara committed:
more

pavelsavara committed:
- remove FeatureSwitchDefinition for now

- add tracing

pavelsavara committed:
whitespace

pavelsavara committed:
navigation draft

pavelsavara committed:
cleanup

pavelsavara committed:
cleanup

pavelsavara committed:
more

pavelsavara committed:
IsAllDataRequested

pavelsavara commented:
> Having many low level metrics could cause performance issues.

I removed few and left on the most useful ones. 
The only `aspnetcore.components.parameters.duration` is per Blazor component. 
It's async and  executes customer's business logic. 
If they have thousands of them they are in trouble anyway. And this will help them to figure it out.
The rest of them are per request, which should be OK.

I have 2 remaining issues
- on metrics/duration histogram I'm unable to see any time bigger than `5` in dashboard. I assume it is the 0.005s bucket. Even if I have added 1s delay and validated stopwatch said 1.1s TotalSeconds. I think there is some problem with aggregation or display ? Or maybe all the other fast examples are drowning this one ?
- on tracing/activities I would like to link HTTP activity from the place where SignalR is creating the Blazor circuit. So I capture it's `Activity.Current.Context` and use it later to `AddLink()` on my activity. In some cases it leads to the HTTP activity, but in may cases the HTTP activity is not in the dashboard at all. I'm thinking it could be sampling. I would like to skip the link if I know that the HTTP activity was not selected for sampling. But HTTP activity always (on dev machine without pressure) have `.Recorded` true and sometimes is missing on the dashboard anyway. 

I would appreciate hints, many thanks! @noahfalk @JamesNK

BrennanConroy commented:
> * I would like to link HTTP activity from the place where SignalR is creating the Blazor circuit. So I capture it's `Activity.Current.Context` and use it later to `AddLink()` on my activity.

I don't know how Blazor circuits are created, but if it's from a Hub method then Activity.Current won't be the HTTP activity. We hop off the HTTP activity on purpose in SignalR: https://github.com/dotnet/aspnetcore/blob/9f2b0880ff3c09db4a8915d62788a7d1ba62df64/src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs#L398-L403

> * in may cases the HTTP activity is not in the dashboard at all

Is that because the HTTP request is still running? I don't think activites show up in the dashboard until they're stopped, and if you're using SignalR you're likely using a websocket request which is long running.

pavelsavara commented:
> I don't know how Blazor circuits are created, but if it's from a Hub method then Activity.Current won't be the HTTP activity. We hop off the HTTP activity on purpose in SignalR:

I'm capturing `Activity.Current.Context` in `ComponentHub` constructor, which happens before any `Activity` changes in the SignalR layer, and so I'm able to capture HTTP Activity. Capture works, I test the `Name` and I get the `TraceId` just fine.

> Is that because the HTTP request is still running? I don't think activites show up in the dashboard until they're stopped, and if you're using SignalR you're likely using a websocket request which is long running.

This is it, thank you @BrennanConroy !

pavelsavara commented:
It's also topic to discuss for long running activities on Blazor. 
- current circuit - this is object in memory, with SignalR connection. 
- current route - this is mo logical than physical. But it makes sense to link click event activities to the current route

We have 2 way how to deal with them I think
- or keep them open for the whole duration
    - should we install them back to `Activity.Current` on next HTTP/SignalR request for child Activity sake ?
    - Should we use parent relationship instead of link ? I understood the feedback in [29846](https://github.com/dotnet/aspnetcore/issues/29846) to mean that long running parent makes bad UX in the dashboard.
- close them soon, in context of the current event/click/navigation
    - this makes them visible in dashboard/OTEL and linkable
    - but we don't capture the true duration other sub-spans: disconnect, reconnect, close

Right now I have short+links implementation. 

I guess developers use OTEL mostly in production and so even the long running traces would be recorded already.

But maybe developers also use it in dev loop ? In which case it would be great to have "trace preview" for thing that started but not stopped yet. To not get confused the same way as I did.

pavelsavara committed:
- ComponentsActivitySourceTest

- cleanup

samsp-msft commented:
Adding a general naming one here - `Microsoft.AspNetCore.Components` I would have no idea that this was for Blazor. I can understand why you may not want to directly put Blazor in the name, but maybe `Microsoft.AspNetCore.UI` or `Microsoft.AspNetCore.UIComponents` would be more obvious?

samsp-msft commented on src/Components/Components/src/ComponentsMetrics.cs:
```diff
@@ -0,0 +1,235 @@
+            description: "Total number of exceptions during browser event processing.");
+
+        _parametersDuration = _meter.CreateHistogram(
+            "aspnetcore.components.parameters.duration",
```
As a neophyte Blazor developer, I don't quite understand why you'd want metrics broken down to the level of parameters. I am probably misunderstanding what this represents?

noahfalk replied:
Ideally an experienced Blazor developer should be able to make a good guess at the meaning of the metric given just its name. I'm fine to assume those devs understand the meaning of 'parameters'. As described above it sounded like 'parameters' is a noun that doesn't inherently have a notion of time duration associated with it? Perhaps we could name this `aspnetcore.components.update_parameters.duration` to emphasize the update is the time duration that is being measured?

pavelsavara replied:
> `aspnetcore.components.update_parameters.duration`

Sounds good to me.

samsp-msft replied:
I think the granularity here is probably too fine. I don't think we need to be tracking this on a per component/parameter basis - there could potentially be hundreds of those on a page. I would suggest that we focus on what the end user will see which is that they take an action and that results in an update to the page. That will admittedly include a network round-trip, but understanding it from the server level is probably sufficient as it is what is in the developers control (unlike the network from the browser)

pavelsavara replied:
The user code that's running in the triggered events could make async HTTP call or database calls. If they do SELECT N+1 anti-pattern, it would be visible here. 

Those problems are currently not easy to diagnose, especially if the components are from different vendors or teams.

I think it's good to know which component was rendered when state changed. How many times and how long it took.

The action they could take based on this data, is to cache/redesign data acquisition in their components or reduce number of components or depth. Maybe also reduce percentage of cases that the sub-tree is re-rendered on data propagation.

Maybe we could have separate meter called `Microsoft.AspNetCore.Components.Lifecycle` which have this finer granularity and `Microsoft.AspNetCore.Components` could be for the big events.

@danroth27 thoughts ?

pavelsavara commented:
> Adding a general naming one here - `Microsoft.AspNetCore.Components` I would have no idea that this was for Blazor.

`Microsoft.AspNetCore.Components` is a C# namespace in which whole Blazor lives since forever.

noahfalk commented on src/Components/Components/src/ComponentsMetrics.cs:
```diff
@@ -0,0 +1,235 @@
+        _meter = meterFactory.Create(MeterName);
+
+        _navigationCount = _meter.CreateCounter<long>(
+            "aspnetcore.components.navigation.count",
```
According to OTel naming guidelines `count` only gets used for UpDownCounters (https://opentelemetry.io/docs/specs/semconv/general/naming/#use-count-instead-of-pluralization-for-updowncounters). I don't know the details on how Blazor routing compares to ASP.NET routing but the prior art from ASP.NET is https://opentelemetry.io/docs/specs/semconv/dotnet/dotnet-aspnetcore-metrics/#metric-aspnetcoreroutingmatch_attempts. Does the action being counted have a similar concept and naming to ASP.NET 'match attempts' or does Blazor do it differently?

noahfalk commented on src/Components/Components/src/ComponentsMetrics.cs:
```diff
@@ -0,0 +1,235 @@
+
+        _navigationCount = _meter.CreateCounter<long>(
+            "aspnetcore.components.navigation.count",
+            unit: "{exceptions}",
```
Copy/paste error?

noahfalk commented on src/Components/Components/src/ComponentsMetrics.cs:
```diff
@@ -0,0 +1,235 @@
+            advice: new InstrumentAdvice<double> { HistogramBucketBoundaries = MetricsConstants.ShortSecondsBucketBoundaries });
+
+        _eventException = _meter.CreateCounter<long>(
+            "aspnetcore.components.event.exception",
```
```suggestion
            "aspnetcore.components.event.exceptions",
```
(https://opentelemetry.io/docs/specs/semconv/general/naming/#pluralization)

noahfalk commented on src/Components/Components/src/ComponentsMetrics.cs:
```diff
@@ -0,0 +1,235 @@
+
+        _eventException = _meter.CreateCounter<long>(
+            "aspnetcore.components.event.exception",
+            unit: "{exceptions}",
```
```suggestion
            unit: "{exception}",
```
I think OTel had guidance that unit names should be singular. I can't find it now but that is the pattern all our other metric units follow.

noahfalk commented on src/Components/Components/src/ComponentsMetrics.cs:
```diff
@@ -0,0 +1,235 @@
+            advice: new InstrumentAdvice<double> { HistogramBucketBoundaries = MetricsConstants.ShortSecondsBucketBoundaries });
+
+        _parametersException = _meter.CreateCounter<long>(
+            "aspnetcore.components.parameters.exception",
```
```suggestion
            "aspnetcore.components.parameters.exceptions",
```
If the naming changes for duration it might flow here too. It feels a bit unclear what a 'parameter exception' is, but hopefully a 'parameter update exception' would be well understood.

noahfalk commented on src/Components/Components/src/ComponentsMetrics.cs:
```diff
@@ -0,0 +1,235 @@
+
+        _parametersException = _meter.CreateCounter<long>(
+            "aspnetcore.components.parameters.exception",
+            unit: "{exceptions}",
```
```suggestion
            unit: "{exception}",
```

noahfalk commented on src/Components/Components/src/ComponentsMetrics.cs:
```diff
@@ -0,0 +1,235 @@
+            advice: new InstrumentAdvice<double> { HistogramBucketBoundaries = MetricsConstants.ShortSecondsBucketBoundaries });
+
+        _batchException = _meter.CreateCounter<long>(
+            "aspnetcore.components.rendering.batch.exception",
```
```suggestion
            "aspnetcore.components.rendering.batch.exceptions",
```

noahfalk commented on src/Components/Components/src/ComponentsMetrics.cs:
```diff
@@ -0,0 +1,235 @@
+
+        _batchException = _meter.CreateCounter<long>(
+            "aspnetcore.components.rendering.batch.exception",
+            unit: "{exceptions}",
```
```suggestion
            unit: "{exception}",
```

noahfalk commented:
> Microsoft.AspNetCore.Components.OnCircuit: CIRCUIT {circuitId}
tags: circuit.id
links: HTTP activity
Microsoft.AspNetCore.Components.OnRoute: ROUTE {route} -> {componentType}
tags: circuit.id, component.type, route
links: HTTP trace, circuit trace
Microsoft.AspNetCore.Components.OnEvent: EVENT {attributeName} -> {componentType}.{methodName}
tags: circuit.id, component.type, component.method, attribute.name
links: HTTP trace, circuit trace, router trace

I'm a bit confused by the distributed tracing part of this. Partly that might be my limited experience with Blazor. I'm not sure what period of time is being measured by the different OnCircuit, OnRoute, OnEvent spans. For example I know what a Blazor circuit is but I don't know what 'OnCircuit' is measuring. Is this a span that measures the entire duration of 1 blazor circuit?

I'll probably have more questions once I understand what each of the spans represents.

danroth27 commented:
> aspnetcore.components.rendering.batch.duration - Duration of rendering batch.

I don't think we currently define in our public Blazor docs what a render batch is. The only public mention of render batches that I could find is the [https://learn.microsoft.com//dotnet/api/microsoft.aspnetcore.components.server.circuitoptions.maxbufferedunacknowledgedrenderbatches](CircuitOptions.MaxBufferedUnacknowledgedRenderBatches) property. So, it's not clear to me what this value represents or whether it is useful. Should we be measuring something else that is more directly correlated the publicly documented [component lifecycle](https://learn.microsoft.com/aspnet/core/blazor/components/lifecycle)? Or do we want to introduce the concept of a render batch in our docs?

> aspnetcore.components.rendering.batch.exception - Total number of exceptions during batch rendering.

Again, since render batch isn't currently a publicly defined concept, should we be counting the number of exceptions per some other period?

> aspnetcore.components.event.duration - Duration of processing browser event asynchronously.

I assume this is an average duration of all browser event handlers across the entire app regardless of render mode. That seems reasonable as a high-level view of the responsiveness of the app. But what does the "asynchronously" imply? Are synchronous event handlers not included in this metric?

> aspnetcore.components.parameters.duration - Duration of processing component parameters asynchronously.

I'm not sure what's included in the "processing" of component parameters. Is this the duration of the `OnParametersSet` component lifecycle event?

aspnetcore.components.navigation.count - Total number of route changes.

Is this a total count of all page navigations across the entire app regardless of render mode? What would that be used for?

noahfalk commented:
> What would that be used for?

I was under the (perhaps mistaken?) impression that the scenarios using the metrics had already been looked at and appropriate metrics identified. If not, perhaps a good starting point is to identify what diagnostic questions we'd like users to be able to solve here. Usually I'd recommend:

- P0 - Is my service healthy? (Most responses are successful/correct, latency is reasonable)
- P1 - What is the load on my service? (How many requests/sec?)
- P1 - How many resources is my service using? (Often CPU, memory, disk are already measured elsewhere, but maybe a service would have some specialized resource it consumes that needs to be measured?)
- P2 - Why is my service unhealthy? Metrics can be used to narrow down or even root cause problems but beware the slippery slope of adding too many metrics that are only useful in ever more niche cases. Its not unreasonable to require logs or distributed tracing for problem diagnosis.

pavelsavara committed:
Update src/Components/Components/src/ComponentsMetrics.cs

Co-authored-by: Noah Falk <noahfalk@users.noreply.github.com>

pavelsavara committed:
Update src/Components/Components/src/ComponentsMetrics.cs

Co-authored-by: Noah Falk <noahfalk@users.noreply.github.com>

pavelsavara committed:
Update src/Components/Components/src/ComponentsMetrics.cs

Co-authored-by: Noah Falk <noahfalk@users.noreply.github.com>

pavelsavara commented:
P0-P2 - This is useful angle, thanks!

> P0 - Is my service healthy? (Most responses are successful/correct, latency is reasonable)
- `aspnetcore.components.event.exceptions` - click to which component causes what type of exception. Stats, not trace, this doesn't tell you which sub component failed.
- `aspnetcore.components.event.duration` - click to which component is slow ?

> P1 - What is the load on my service? (How many requests/sec?)
- `aspnetcore.components.circuits.count` - how many sessions I processed today ?
- `aspnetcore.components.event.duration` - this is also counter, how many clicks I processed today ?
- `aspnetcore.components.circuits.duration` - this is session duration, interesting for many other KPIs
- `aspnetcore.components.navigation.count` - how many different Blazor pages people visited ?

> P1 - How many resources is my service using?
- `aspnetcore.components.circuits.active_circuits` - proxy to how much memory the sessions state holds ?
- `aspnetcore.components.circuits.connected_circuits` - how many signalR connections are open ? 
- Delta between the two above is about WebSocket disconnect/re-connect, network quality, browser tab going to sleep etc.

> P2 - Why is my service unhealthy?
- `aspnetcore.components.parameters.exceptions` - which component failed ? Exception type tag/stats, no trace.
- `aspnetcore.components.rendering.batch.exceptions` - which component failed ?
- `aspnetcore.components.parameters.duration` - which component makes my app rendering slow ?
- `aspnetcore.components.rendering.batch.duration` - which component makes events slow ? Also how many UI elements were in the diff ?

Note, I also mention `aspnetcore.components.circuits` metrics which already landed before, but we can improve them too if you want.

pavelsavara commented:
> For example I know what a Blazor circuit is but I don't know what 'OnCircuit' is measuring. Is this a span that measures the entire duration of 1 blazor circuit?

We can definitely improve naming. 

This goes back to my [questions about long running activities](https://github.com/dotnet/aspnetcore/pull/61609#issuecomment-2834256951). 
That activity is relatively short lived at the moment, compared to whole life-time of the circuit. We can change that with consequences for inner dev loop.

`OnCircuit` - is representing logical circuit (duration), but at the moment, we stop it earlier.
`OnRoute` - is representing logical route/page in the app. It links to circuit (when Blazor interactive).
`OnEvent` - in Blazor interactive, something was clicked. It typically happens inside of specific route and circuit which are linked.

pavelsavara commented:
> > aspnetcore.components.rendering.batch.duration - Duration of rendering batch.
> 
> I don't think we currently define in our public Blazor docs what a render batch is. 

Maybe we just need to rename it? Anyway, this is more on the troubleshooting side of misbehaving component. Producing long diffs/batches leads to network traffic, latency and slow rendering. 

As I suggested above, we could have separate namespace for it with separate opt-in.
Or we could drop it and circle back in the future.

> > aspnetcore.components.rendering.batch.exception - Total number of exceptions during batch rendering.
> should we be counting the number of exceptions per some other period?

We also count exceptions per click/event. But I need to see if the exceptions from batch related problems would appear there.

> > aspnetcore.components.event.duration - Duration of processing browser event asynchronously.
> 
> I assume this is an average duration of all browser event handlers across the entire app regardless of render mode. 

At the moment this works only for SignalR interactive. I think we could also make it work for form-submit.
Making it work for WASM means that we need to fix OTEL for WASM and implement some publishing for it. 
It's out of scope for Net10.

> what does the "asynchronously" imply?

I already renamed this and dropped "async". It means including your DB request or whatever async business logic.

> Is this the duration of the `OnParametersSet` component lifecycle event?

Yes, or `OnInitialized`.

> > aspnetcore.components.navigation.count - Total number of route changes.
> 
> Is this a total count of all page navigations across the entire app regardless of render mode? What would that be used for?

Except WASM. It has the route pattern as tag/dimension that you can use as filter. It's more business oriented KPI. Which of my pages are hot ?

pavelsavara committed:
feedback

pavelsavara commented:
Making circuit/route activity/trace them long lived has troubles with re-installing them into `Activity.Current`. 
And then troubles display of hours long one with sub-spans in the UI.
So, I think it's little benefit.

If we keep them short, maybe they should be literally 0ms long. Just an context anchor, grouping other traces.

Re Activity names: they are not very visible in the Aspire UI, and `DisplayName` prevails.

Circuit Activity/trace is created in internal `ComponentHub.StartCircuit` and so maybe `StartCircuit` is good name. There is also public API `CircuitHandler.OnCircuitOpenedAsync`, but I dislike `Async` postfix and also `OnCircuitOpenedAsync` happens short while later, being triggered by another message from JS. But `OnCircuitOpened` would be my second choice.

Route circuit is created in `Router.SetParametersAsync` -> `Router.Refresh`. Not great names for the activity.
Maybe we can change it to `OnRouteChanged`

Regarding click/event. We already have concept of [event](https://learn.microsoft.com/en-us/aspnet/core/blazor/components/event-handling). The activity should be active thru whole duration of `DispatchEventAsync` and would be parent for any business logic and distributed HTTP client traces. 

I would like event Activity also trigger for form submit, interop call from JS, and enhanced navigation.

Maybe we can change it to `Event` or `BrowserEvent` ?

pavelsavara committed:
fix tests

pavelsavara committed:
update_parameters feedback

pavelsavara commented:
Sam: Is circuitId useful for DisplayName of Circuit activity? What else we could display there instead. Could we have IP address ?

samsp-msft commented:
I met with @pavelsavara and I now understand what everything is for - it looks great. I think customers will be really happy with this. My concerns about granularity in terms of sending too much data have been mitigated.

pavelsavara commented:
But I think this problem of mapping traces to users in hotline is not specific to Blazor.

I'm linking the HTTP Activity/trace that created the circuit, so if there is more tags on the HTTP trace, they could use that (after the session and the long running HTTP/WS connection is finished).

Sometimes there is authenticated user in the HTTP context, but I think it's probably not good for use to expose any PII.

If the app developer wanted to add more tags, I believe that they could capture `Activity.Current` and add something.

Maybe the app developer could also add "show my circuitID" into application settings menu and the IT call center could ask them to click it. The circuitID is random number and it's not a secret from security perspective.

samsp-msft commented:
> If the app developer wanted to add more tags, I believe that they could capture `Activity.Current` and add something.

The activity is created in CircuitFactory at line 127 and is stopped at line 173. Is there any user code executed between these points where the activity would be active, so they can retrieve it and add tags. Once its stopped, AFAIK its too late to add anything to it.

samsp-msft commented on src/Components/Components/src/ComponentsActivitySource.cs:
```diff
@@ -0,0 +1,177 @@
+                }
+            }
+
+            activity.DisplayName = $"ROUTE {route ?? "unknown"} -> {componentType ?? "unknown"}";
```
```suggestion
            activity.DisplayName = $"Route {route ?? "[unknown path]"} -> {componentType ?? "[unknown component]"}";
```

samsp-msft commented on src/Components/Components/src/ComponentsActivitySource.cs:
```diff
@@ -0,0 +1,177 @@
+                }
+            }
+
+            activity.DisplayName = $"EVENT {attributeName ?? "unknown"} -> {componentType ?? "unknown"}.{methodName ?? "unknown"}";
```
```suggestion
            activity.DisplayName = $"Event {attributeName ?? "[unknown]"} -> {componentType ?? "[unknown]"}.{methodName ?? "[unknown]"}";
```

samsp-msft commented on src/Components/Components/src/ComponentsMetrics.cs:
```diff
@@ -0,0 +1,239 @@
+            "aspnetcore.components.update_parameters.duration",
+            unit: "s",
+            description: "Duration of processing component parameters.",
+            advice: new InstrumentAdvice<double> { HistogramBucketBoundaries = MetricsConstants.ShortSecondsBucketBoundaries });
```
`ShortSecondsBucketBoundaries` is designed for network level latencies such as an http request. I suspect that components may need buckets more focused on the sub ms level timings?

pavelsavara replied:
This could involve network requests or not. 
Added `BlazorPropertyUpdateSecondsBucketBoundaries = [0.000001, 0.00001, 0.0001, 0.001, 0.005, 0.01, 0.025, 0.05, 0.075, 0.1, 0.25, 0.5, 0.75, 1, 2.5, 5, 7.5, 10]`

samsp-msft commented on src/Components/Components/src/ComponentsMetrics.cs:
```diff
@@ -0,0 +1,239 @@
+            "aspnetcore.components.rendering.batch.duration",
+            unit: "s",
+            description: "Duration of rendering batch.",
+            advice: new InstrumentAdvice<double> { HistogramBucketBoundaries = MetricsConstants.ShortSecondsBucketBoundaries });
```
same as above for the durations

pavelsavara replied:
Added `VeryShortSecondsBucketBoundaries = [0.000001, 0.00001, 0.0001, 0.001, 0.01, 0.1, 1]`

samsp-msft commented on src/Components/Components/src/ComponentsMetrics.cs:
```diff
@@ -0,0 +1,239 @@
+        var tags = new TagList
+        {
+            { "component.type", componentType ?? "unknown" },
+            { "route", route ?? "unknown" },
```
should `route` be `http.route` to match the http metrics / semconv? https://learn.microsoft.com/en-us/aspnet/core/log-mon/metrics/built-in?view=aspnetcore-9.0#metric-httpserverrequestduration

pavelsavara replied:
This is route in terms of browser address bar URL segment. But it is not always HTTP.
I don't have strong opinion either way.

@noahfalk thoughts ?

samsp-msft commented on src/Components/Components/src/ComponentsMetrics.cs:
```diff
@@ -0,0 +1,239 @@
+        var tags = new TagList
+        {
+            { "component.type", componentType ?? "unknown" },
+            { "route", route ?? "unknown" },
```
if we don't know the route, is it better to skip the tag all together? I can see it going both ways - it makes it more explicit this way.

pavelsavara replied:
I'm thinking about running a query on this. Leaving the tag out could mean you need "LEFT JOIN" to not lose the whole activity row. I don't know if this is realistic situation.

pavelsavara replied:
the 3rd option is tag is present but value is null.  `{ "route", null },` just for completeness

samsp-msft commented on src/Components/Components/src/ComponentsMetrics.cs:
```diff
@@ -0,0 +1,239 @@
+            {
+                { "component.type", componentType ?? "unknown" },
+                { "component.method", methodName ?? "unknown" },
+                { "attribute.name", attributeName ?? "unknown" }
```
I understand why this is based on the attribute, but its really also the event name. would that make more sense as "event"?

samsp-msft commented on src/Components/Components/src/ComponentsMetrics.cs:
```diff
@@ -0,0 +1,239 @@
+
+            var tags = new TagList
+            {
+                { "component.type", componentType ?? "unknown" },
```
`code.class.name`? to match https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/code.md

pavelsavara replied:
We don't really have code reference, as in file name and line number.

noahfalk replied:
My initial thought is that the thing being refered to was the Blazor component abstraction in which case component.type and component.method made sense to me. The fact that code.class.name isn't currently defined also nudges me in that direction. If we did want to use code.* attributes instead this feels like another spot where OTel semconv feedback could be helpful.

samsp-msft commented on src/Components/Components/src/ComponentsMetrics.cs:
```diff
@@ -0,0 +1,239 @@
+            var tags = new TagList
+            {
+                { "component.type", componentType ?? "unknown" },
+                { "component.method", methodName ?? "unknown" },
```
`code.function.name` is the semconv for this - it may mean we don't need a separate one for the component?

pavelsavara replied:
having separate tag for component could make it easier to filter when running queries on it. "contains" operation makes bad use of DB index on join.

samsp-msft commented on src/Components/Components/src/ComponentsMetrics.cs:
```diff
@@ -0,0 +1,239 @@
+    {
+        var tags = new TagList
+        {
+            { "diff.length.bucket", BucketEditLength(diffLength) }
```
`diff.approx_length`? I was a bit confused by the bucket in the name. Totally the right thing to do to bucketize the sizes as that eliminates dimension explosion, but the name confused me as to whether the feature has buckets or something)

@lmolkova - do you know of a precedent for where this is done elsewhere?

pavelsavara replied:
made it `diff.approximate.length`, happy to change it further

samsp-msft commented on src/Shared/Metrics/MetricsConstants.cs:
```diff
@@ -11,6 +11,6 @@ internal static class MetricsConstants
-    // For Blazor/signalR sessions, which can last a long time.
-    public static readonly IReadOnlyList<double> VeryLongSecondsBucketBoundaries = [0.5, 1, 2, 5, 10, 30, 60, 120, 300, 600, 1500, 60*60, 2 * 60 * 60, 4 * 60 * 60];
+    // For blazor circuit sessions, which can last a long time.
+    public static readonly IReadOnlyList<double> VeryLongSecondsBucketBoundaries = [1, 10, 30, 1 * 60, 2 * 60, 3 * 60, 4 * 60, 5 * 60, 6 * 60, 7 * 60, 8 * 60, 9 * 60, 10 * 60, 1 * 60 * 60, 2 * 60 * 60, 3 * 60 * 60, 6 * 60 * 60, 12 * 60 * 60, 24 * 60 * 60];
```
These should ideally be exponential eg
1, 3, 10, 30,
1*60, 3*60, 10*60, 30*60,
1*60*60, 3*60*60, 10*60*60, 24*60*60  

This uses an approx multiplier of 3, while meeting user sensible boundaries

pavelsavara replied:
done

pavelsavara replied:
Also renamed to be Blazor-specific to `BlazorCircuitSecondsBucketBoundaries`

samsp-msft reviewed (commented):
Some naming suggestions.

For sessions, I wonder if when log messages are fired, is the activity context going to be the session activity, and if not is there a way to force it to be? @noahfalk  this is an issue with essentially zero length spans - you might want to force log messages to be parented to it but something else is the activity at that point?

noahfalk commented:
> This goes back to my https://github.com/dotnet/aspnetcore/pull/61609#issuecomment-2834256951.
That activity is relatively short lived at the moment, compared to whole life-time of the circuit. We can change that with consequences for inner dev loop.


> OnCircuit - is representing logical circuit (duration), but at the moment, we stop it earlier.
OnRoute - is representing logical route/page in the app. Logicaly it should be active until you navigate elsewhere, but right now we stop it early. It links to circuit (when Blazor interactive).
OnEvent - in Blazor interactive, something was clicked. It typically happens inside of specific route and circuit which are linked.

> Right now, the short circuit and route activities mostly serve as something that click event activities could link to. For the context.

Thanks, that helps a bunch on understanding where things are at now. A couple thoughts:

- In the question how to handle long-lived activities I think there is a 3rd option which is not to produce the activity at all. Instead of an event linking to the activity the only reference would be the circuit.id tag or route.id tag.
- Using links and ~0-length activities as a grouping/linking mechanism feels like a novel use of distributed tracing telemetry that many tools may not visualize in a helpful way. Have you tried looking at any of the telemetry in Azure Monitor's portal? As I recall Azure Monitor's viewer is quite aggressive at following links transitively in both the backwards and forwards directions. I'm suspicious that if you try to visualize any event AzureMonitor is going to wind up showing you every route and event in the entire circuit superimposed on a single very cluttered timeline.
- If we keep Circuit and Route I agree a name like CircuitStart or RouteStart/RouteChanged might be more readily understood. For Event perhaps we drop the 'On' and just name it 'Event' or 'HandleEvent'?

> For sessions, I wonder if when log messages are fired, is the activity context going to be the session activity, and if not is there a way to force it to be? @noahfalk this is an issue with essentially zero length spans - you might want to force log messages to be parented to it but something else is the activity at that point?

Activity.Current can be set at any time, but it can't be set to an Activity that is already stopped. If you did want to produce the set of all log messages that occurred within a given circuit I think the best options would be one of:
- In the telemetry viewer, query for all activities where circuit.id tag = X, then merge all logs linked to those activities. (This requires tools to support a table JOIN across logs and distributed traces which probably isn't universally supported)
- Add circuit.id as another field in the log messages and query for it directly. circuit.id could either be an explicit field in select log messages, or it could be added via enrichment. Once the field is there querying for all logs where circuit.id=X would return the desired logs.

pavelsavara committed:
Update src/Components/Components/src/ComponentsActivitySource.cs

Co-authored-by: Sam Spencer <54915162+samsp-msft@users.noreply.github.com>

pavelsavara committed:
Update src/Components/Components/src/ComponentsActivitySource.cs

Co-authored-by: Sam Spencer <54915162+samsp-msft@users.noreply.github.com>

pavelsavara committed:
Update src/Components/Components/src/ComponentsActivitySource.cs

Co-authored-by: Sam Spencer <54915162+samsp-msft@users.noreply.github.com>

pavelsavara committed:
feedback

pavelsavara committed:
ActivityKind.Internal

pavelsavara commented on src/Components/Components/src/ComponentsActivitySource.cs:
```diff
@@ -0,0 +1,177 @@
+            _httpContext = CaptureHttpContext();
+        }
+
+        var activity = ActivitySource.CreateActivity(OnRouteName, ActivityKind.Server, parentId: null, null, null);
```
done

pavelsavara committed:
aspnetcore.components prefix for tags

pavelsavara commented:
All, I created https://github.com/open-telemetry/semantic-conventions/pull/2235 in order to validate and document the new metrics in OTEL community.

As a result I updated all tags in this PR with prefix `aspnetcore.components`

pavelsavara commented:
@lmolkova's feedback on the other PR is

- merge duration and exception metrics into single histogram metric. Our HTTP _requestDuration already does that too. I will do it.

- rename `aspnetcore.components` metrics prefix to -> `aspnetcore.component` because of otel pluralization rules. 
This would create mismatch with C# existing namespace. I prefer C# namespace because that is a existing namespace already. I will keep it as it is, unless somebody here convinces me otherwise.

- "event" means something specific in otel. But also something in the browser and in Blazor. Suggestion to rename `aspnetcore.components.event` to `aspnetcore.components.operation` or `aspnetcore.components.call`. Again, I prefer to stick to Blazor meaning.

pavelsavara committed:
- merge exception and duration metrics

- rename aspnetcore.components.circuit to aspnetcore.components.circuits
- rename circuits.active_circuits to circuit.active and connected_circuits to circuit.connected
- rename aspnetcore.components.event.duration to aspnetcore.components.event_handler and include exceptions
- rename aspnetcore.components.update_parameters.duration to aspnetcore.components.update_parameters and include exceptions
- rename aspnetcore.components.rendering.batch.duration to aspnetcore.components.render_diff and include exceptions

pavelsavara committed:
Event -> HandleEvent  trace rename

pavelsavara committed:
remove aspnetcore.components.circuit.count

pavelsavara committed:
Improve descriptions

javiercn commented on src/Components/Server/src/Circuits/RemoteRenderer.cs:
```diff
@@ -369,7 +369,7 @@ private async Task CaptureAsyncExceptions(Task task)
     }
 
-    private static partial class Log
+    private static new partial class Log
```
Is this `new` really needed here?

pavelsavara replied:
It's a warning now after I added the project reference.
It is hiding the parent logger. 

A) If we want to use the parent logger, I can delete this one.
B) or I can #pragma the warning

javiercn commented on src/Components/Components/src/RegisterRenderingMetricsServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,57 @@
+    public static IServiceCollection AddComponentsMetrics(
+        IServiceCollection services)
+    {
+        // do not register IConfigureOptions<StartupValidatorOptions> multiple times
+        if (!IsMeterFactoryRegistered(services))
+        {
+            services.AddMetrics();
+        }
```
Is this check to avoid calling `services.AddMetrics()` more than once?

javiercn commented on src/Components/Server/src/Circuits/CircuitHost.cs:
```diff
@@ -51,6 +52,7 @@ public CircuitHost(
         RemoteNavigationManager navigationManager,
         CircuitHandler[] circuitHandlers,
         CircuitMetrics? circuitMetrics,
+        ComponentsActivitySource? componentsActivitySource,
```
The activity source and the metrics shouldn't be nullable, they should be the same as RemoteJsRuntime and other parameters.

We unconditionally register them https://github.com/dotnet/aspnetcore/pull/61609/files#diff-e82512017622b574554e1a1519f2bb3bbb1d8957e5346b3a4cbeb0f10e368c79R64 and

https://github.com/dotnet/aspnetcore/pull/61609/files#diff-9e461d4480403d04ed6d5e6c31c44261fe8d5926969ab338dd910564ea211396R79-R80

We should throw an exception if they aren't being provided.

javiercn commented on src/Components/Components/src/ComponentsActivitySource.cs:
```diff
@@ -0,0 +1,177 @@
+
+    private ActivitySource ActivitySource { get; } = new ActivitySource(Name);
+
+    public static ActivityContext CaptureHttpContext()
+    {
+        var parentActivity = Activity.Current;
+        if (parentActivity is not null && parentActivity.OperationName == "Microsoft.AspNetCore.Hosting.HttpRequestIn" && parentActivity.Recorded)
+        {
+            return parentActivity.Context;
+        }
+        return default;
+    }
```
Layering wise this shouldn't go here.

`Microsoft.AspNetCore.Components` shoudn't have a dependency on `Microsoft.AspNetCore.Hosting`

pavelsavara replied:
Please be more specific, what do you want to see instead ?

JamesNK replied:
Are you trying to get the activity of the current request here? There is a feature in the features collection for doing that: https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.features.ihttpactivityfeature.activity?view=aspnetcore-9.0

pavelsavara replied:
Can I get that from DI without passing HttpContext thru all SignalR layers ? Or from some TLS ?

JamesNK replied:
It comes from HttpContext.

javiercn commented on src/Components/Components/src/ComponentsActivitySource.cs:
```diff
@@ -0,0 +1,177 @@
+    public Activity? StartRouteActivity(string componentType, string route)
+    {
+        if (_httpContext == default)
+        {
+            _httpContext = CaptureHttpContext();
+        }
+
+        var activity = ActivitySource.CreateActivity(OnRouteName, ActivityKind.Internal, parentId: null, null, null);
+        if (activity is not null)
+        {
+            if (activity.IsAllDataRequested)
+            {
+                if (_circuitId != null)
+                {
+                    activity.SetTag("aspnetcore.components.circuit.id", _circuitId);
+                }
+                if (componentType != null)
+                {
+                    activity.SetTag("aspnetcore.components.type", componentType);
+                }
+                if (route != null)
+                {
+                    activity.SetTag("aspnetcore.components.route", route);
+                }
+                if (_httpContext != default)
+                {
+                    activity.AddLink(new ActivityLink(_httpContext));
+                }
+                if (_circuitContext != default)
+                {
+                    activity.AddLink(new ActivityLink(_circuitContext));
+                }
+            }
+
+            activity.DisplayName = $"Route {route ?? "[unknown path]"} -> {componentType ?? "[unknown component]"}";
+            activity.Start();
+            _routeContext = activity.Context;
+        }
+        return activity;
+    }
```
This should go on the Server assembly.

pavelsavara replied:
If we move it to different class/instance, we would not be able to use `_circuitId` and `_circuitContext`. And we would have to create lot of semi-public interfaces to pass that info between those classes. Are you sure it's worth it ?

javiercn replied:
Isn't the CircuitId on a parent activity? (like start circuit or something like that)

All the `_httpContext`, `_circuitContext` on that event is at the wrong level of abstraction. I don't know if it can be set on the parent activity and flowed into this one, but the route activity should not know about _circuitContext or _httpContext.

To put it in a different way, this method/class should not be responsible for resolving the parent activity to link to. That should be provided somehow. We might need public API for that.

We don't want to keep adding to this list when/if we add new render modes. For example, I imagine WebView migth also want to track its own parent activity and it would require us to add it here.

pavelsavara replied:
Let's address this in next PR.

I agree that we should not break existing expectations about layers. On the other hand, I would like to keep this as simple as possible. This is internal class and I don't prefer lofty layered architectures for simple and perf sensitive classes like this one.

javiercn commented on src/Components/Components/src/ComponentsActivitySource.cs:
```diff
@@ -0,0 +1,177 @@
+                }
+                if (_circuitContext != default)
+                {
+                    activity.AddLink(new ActivityLink(_circuitContext));
+                }
+                if (_routeContext != default)
+                {
+                    activity.AddLink(new ActivityLink(_routeContext));
+                }
+            }
+
+            activity.DisplayName = $"Event {attributeName ?? "[unknown attribute]"} -> {componentType ?? "[unknown component]"}.{methodName ?? "[unknown method]"}";
+            activity.Start();
+        }
+        return activity;
+    }
+
+    public static void FailEventActivity(Activity? activity, Exception ex)
+    {
+        if (activity != null && !activity.IsStopped)
+        {
+            activity.SetTag("error.type", ex.GetType().FullName);
+            activity.SetStatus(ActivityStatusCode.Error);
+            activity.Stop();
+        }
+    }
+
+    public static async Task CaptureEventStopAsync(Task task, Activity? activity)
+    {
+        try
+        {
+            await task;
+            activity?.Stop();
+        }
+        catch (Exception ex)
+        {
+            FailEventActivity(activity, ex);
+        }
+    }
+}
```
This probably needs to go into a different class `ComponentEventActivitySource` (or any other name, RendererEventMetrics or something like that)

javiercn commented on src/Components/Components/src/RenderTree/Renderer.cs:
```diff
@@ -770,6 +813,7 @@ private void ProcessRenderQueue()
 
         _isBatchInProgress = true;
         var updateDisplayTask = Task.CompletedTask;
+        var batchStartTimestamp = ComponentMetrics != null && ComponentMetrics.IsBatchEnabled ? Stopwatch.GetTimestamp() : 0;
```
Don't remember the exact syntax, but you can write it like below

```suggestion
        var batchStartTimestamp = ComponentMetrics  is { IsBatchEnabled: true } ? Stopwatch.GetTimestamp() : 0;
```

javiercn commented on src/Components/Components/src/Rendering/ComponentState.cs:
```diff
@@ -51,6 +52,11 @@ public ComponentState(Renderer renderer, int componentId, IComponent component,
+
+        if (_renderer.ComponentMetrics != null && _renderer.ComponentMetrics.IsParametersEnabled)
+        {
+            _componentTypeName = component.GetType().FullName;
```
Is this perf related or "convenience" related.

Adding a field to ComponentState is not great, since this increments the cost of each component that gets created (which your app can have thousands of).

pavelsavara replied:
This is for perf. Because we need that type _componentTypeName on all parameter updates on each component instance. (when this metric is enabled)

pavelsavara committed:
Merge branch 'main' into blazor_metrics_feedback

pavelsavara commented:
I will merge this so that we could demo the feature soon. There are few pending issues, which I will address on next PR.
